src/android/Dgram.java

```diff
@@ -75,198 +48,191 @@ public void run() {
                         Log.d(TAG, "Exiting message loop because socket is closed.");
                         return;
                     }
 
                     // Reset the length in case we receive an incomplete DatagramPacket
                     datagramPacket.setLength(data.length);
-                    this.m_datagramSocket.receive(datagramPacket);
+                    this.datagramSocket.receive(datagramPacket);
                     String message = new String(data, 0, datagramPacket.getLength(), "UTF-8");
                     String address = datagramPacket.getAddress().getHostAddress();
                     int port = datagramPacket.getPort();
                     Log.d(TAG, "Received message " + message + " from " + address);
-                    Dgram.this.m_javascriptQueue.put("_onMessage("
-                            + this.m_datagramSocketId + ","
-                            + "'" + message + "',"
-                            + "'" + address + "',"
-                            + port + ");");
-                } catch (InterruptedException e) {
-                    Log.d(TAG, e.toString());
-                } catch (Exception e) {
-                    Log.d(TAG, "Receive exception:" + e.toString());
-
-                    try {
-                        Dgram.this.m_javascriptQueue.put("_onError("
-                                + this.m_datagramSocketId
-                                + ",'receive','"
-                                + e.toString() + "');");
-                    } catch (Exception innerException) {
-                        Log.e(TAG, "Receive exception:" + innerException.toString());
-                    } finally {
-                        return;
+                    if (onMessageCallback != null) {
+                        sendMessageResult(message, address, port);
                     }
+                } catch (Exception e) {
+                    Log.d(TAG, "Received exception:" + e.toString());
+                    sendMessageErrorResult(e);
                 }
             }
         }
     }
 
+    private void sendMessageResult(String message, String address, int port) {
+        JSONObject payload = new JSONObject();
+
+        try {
+            payload.put("message", message);
+            payload.put("address", address);
+            payload.put("port", port);
+            CallbackUtil.sendPluginResult(onMessageCallback, new PluginResult(PluginResult.Status.OK, payload));
+        } catch (Exception e) {
+            PluginResult result = new PluginResult(
+                    PluginResult.Status.ERROR,
+                    "Error occurred while creating onMessage payload: " + e.toString()
+            );
+            CallbackUtil.sendPluginResult(onMessageCallback, result);
+        }
+    }
+
+    private void sendMessageErrorResult(Exception e) {
+        try {
+            JSONObject payload = new JSONObject();
+
+            payload.put("error", e.toString());
```

Comment:
i feel like if we are going to use the `onMessageCallback` before communicating errors we should be consistent in the structure and always pass a json object with with an `error` property.  i like they structure here but in the catch it is just passing a string.  so then the javascript has to interrogate the information to figure out what it is getting and what it should do.  also, it isn't clear to me there is value it trying `CallbackUtil.sendPluginResult(onMessageCallback` again immediately in the catch and it feels like it could actually result in an exception within the catch.

Reply:
note, it the previous implementation `Dgram.this.m_javascriptQueue.put` was a safe operation.  all it did was add a string to a queue for later processing so it could be done in a catch.